pulser-core/pulser/backend/remote.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2023 Pulser Development Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Base classes for remote backend execution."""
+from __future__ import annotations
+
+import typing
+from abc import ABC, abstractmethod
+from enum import Enum, auto
+from typing import Any
+
+from pulser.backend.abc import Backend
+from pulser.devices import Device
+from pulser.result import Result, Results
+from pulser.sequence import Sequence
+
+
+class SubmissionStatus(Enum):
+    """Status of a remote submission."""
+
+    PENDING = auto()
+    RUNNING = auto()
+    DONE = auto()
+    CANCELED = auto()
+    TIMED_OUT = auto()
+    ERROR = auto()
+    PAUSED = auto()
+
+
+class RemoteResultsError(Exception):
+    """Error raised when fetching remote results fails."""
+
+    pass
+
+
+class RemoteResults(Results):
+    """A collection of results obtained through a remote connection.
+
+    Args:
+        submission_id: The ID that identifies the submission linked to
+            the results.
+        connection: The remote connection over which to get the submission's
+            status and fetch the results.
+    """
+
+    def __init__(self, submission_id: str, connection: RemoteConnection):
+        """Instantiates a new collection of remote results."""
+        self._submission_id = submission_id
+        self._connection = connection
+
+    @property
+    def results(self) -> tuple[Result, ...]:
+        """The actual results, obtained after execution is done."""
+        return self._results
+
+    def get_status(self) -> SubmissionStatus:
+        """Gets the status of the remote submission."""
+        return self._connection._get_submission_status(self._submission_id)
+
+    def __getattr__(self, name: str) -> Any:
+        if name == "_results":
+            if self.get_status() == SubmissionStatus.DONE:
+                self._results = tuple(
+                    self._connection._fetch_result(self._submission_id)
+                )
+                return self._results
+            raise RemoteResultsError(
+                "The results are not available. The submission's status is "
+                f"{str(self.get_status())}."
+            )
+        raise AttributeError(
+            f"'RemoteResults' object has no attribute '{name}'."
+        )
+
+
+class RemoteConnection(ABC):
+    """The abstract base class for a remote connection."""
+
+    @abstractmethod
```

Comment:
@HGSilveri this is meant to be general for any kind of remote backend, right? It looks all the ingredients are there in the interface. Why do you make the `_fetch_result` and `_get_status` methods private? I guess they can be part of the public interface since those are methods that will be called frequently, I think.

Reply:
Yes, this should be general. The reason why I made `_fetch_result` and `_get_status` private is because they require handling some sort of submission identifier that I wanted to hide from the user if possible. Bear in mind that they still have access to both these features through `RemoteResults.results` and `RemoteResults.get_status()`, which to me is more natural as this object is what is return by the `run()` call.